### PR TITLE
Add support for generic fold output transforms

### PR DESCRIPTION
**Pull request by manupak**
Add support for generic fold output transforms

This commit adds support for folding the output transforms
of gemm-based operator should there be view like ops
after the gemm-based operators.

Additionally folding of view-like ops is lifted to
FoldTranspose pass from AlignTiling.

closes : https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/740

### CODE REVIEW

**Comment by manupak** [comment]
build on top of  : https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/873, hence putting to draft for now.

**Comment by manupak** [comment]
For exclusive changes in this PR : https://github.com/ROCmSoftwarePlatform/rocMLIR/pull/884/commits/5c2854a2670db72cf7120fe400fb67ca402fd8ac

File: mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp [hunk 1]
```diff
@@ -39,6 +41,199 @@ using namespace mlir;
...
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp laGeneric,
+                                PatternRewriter &rewriter) const override;
+};
+
+// This function will take strides calculated using reassociation of dimensions
+// that are modified by view-like ops and recompute the affine expr from src
+// view to dest view.
+static void getDelinearizedAffineExpr(mlir::ArrayRef<int64_t> strides,
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp]: Nit: no need to `mlir::` here, and probably anywhere for that matter

Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp]: I'm confused by the naming in this function. I'd expect, for instance, a 3x4x5 memref to have `shapes = [3, 4, 5]` and strides equal to `[20, 5, 1]`, so these shapes vs strides comparisons make no sense to me

**Review by krzysz00 — COMMENTED** [review]
*waves from the naming nitpicks zone*

Overall, this seems like a perfectly solid rewrite.

File: mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp (hunk 1 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp]: Except that those are association subset of the shapes and the strides. So unless all the dimensions are being collapsed to 1d tensor, one will not be able to infer one by the other.

I ll add more docs and probably rename shapes --> associatedDimSizes strides --> associatedDimStrides

File: mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp [hunk 2]
```diff
@@ -39,6 +41,207 @@ using namespace mlir;
...
+};
+
+// This function will take strides of the original layout of corresponding
+// dimensions being reassociated and the expected resultant dim sizes of
+// post-reassociation --> then, produce the set of resultant AffineExprs after
+// the reassociation.
+static void getDelinearizedAffineExpr(
+    ArrayRef<int64_t> originalStridesDimsBeingReassociated,
+    ArrayRef<int64_t> postReassociatedDimSizes, Builder &b,
+    unsigned int position, SmallVectorImpl<mlir::AffineExpr> &res) {
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp:63]: The `mlir::` nit

**Review by krzysz00 — APPROVED** [review]
Seems like a fine general fix

File: mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp (hunk 2 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/FoldTranspose.cpp:63]: Ah missed that.. I ll do a follow up on Monday... Taking this in for now coz of potential conflicts as Im out today.

**Pull request merged**